File: mbw/src/androidTest/java/com/mycelium/wallet/activity/ActivityIdlingResource.java
package com.mycelium.wallet.activity;

import android.app.Activity;
import androidx.annotation.NonNull;
import androidx.test.espresso.IdlingResource;
import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry;
import androidx.test.runner.lifecycle.Stage;

import java.util.Collection;
import java.util.Iterator;

public class ActivityIdlingResource implements IdlingResource {

    private final Class<? extends Activity> mAcivity;
    private final String mName;
    private ResourceCallback mResourceCallback;

    public ActivityIdlingResource(@NonNull Class<? extends Activity> activity) {
        mAcivity = activity;
        mName = "activity " + activity.getName() + "(@" + System.identityHashCode(activity) + ")";
    }

    @Override
    public String getName() {
        return mName;
    }

    @Override
    public boolean isIdleNow() {
        final boolean isIdle = mAcivity != getActivityInstance().getClass();
        if (isIdle && mResourceCallback != null) {
            mResourceCallback.onTransitionToIdle();
        }
        return isIdle;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback resourceCallback) {
        mResourceCallback = resourceCallback;
    }

    private Activity getActivityInstance() {
        final Activity[] mActivity = new Activity[1];
        final Collection<Activity> resumedActivities =
            ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(Stage.RESUMED);
        Iterator<Activity> iterator = resumedActivities.iterator();
        if (iterator.hasNext()) {
            mActivity[0] = iterator.next();
        }
        return mActivity[0];
    }
}

File: mbw/src/androidTest/java/com/mycelium/wallet/activity/ActivityIdlingResource.kt
package com.mycelium.wallet.activity

import android.app.Activity
import androidx.test.espresso.IdlingResource
import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry
import androidx.test.runner.lifecycle.Stage

class ActivityIdlingResource(private val activity: Class<out Activity>) : IdlingResource {
    private val name = "activity " + activity.name + "(@" + System.identityHashCode(activity) + ")"

    private var resourceCallback: IdlingResource.ResourceCallback? = null

    private val activityInstance: Activity?
        get() = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(Stage.RESUMED)
                .firstOrNull()

    override fun getName(): String = name

    override fun isIdleNow(): Boolean {
        if (activityInstance == null || resourceCallback == null){
            return false
        }
        val isIdle = activity != activityInstance!!.javaClass
        if (isIdle && resourceCallback != null) {
            resourceCallback!!.onTransitionToIdle()
        }
        return isIdle
    }

    override fun registerIdleTransitionCallback(resourceCallback: IdlingResource.ResourceCallback) {
        this.resourceCallback = resourceCallback
    }
}